### PR TITLE
TestContext tests and more refactoring

### PR DESCRIPTION
**Pull request by karanjitsingh**
- _package.json_
  - Downgrade typescript since source maps weren't matching
  - use `--noImplicitUseStrict` for building tests
- _src/Constants.ts_
  - Commonplace for all string constants
  - Use extend `String` object with `String.format`
- _src/Context/Attachment.ts_
  - Move `getTestGuid` to `TestContext.ts`
  - Use Constants
- _src/Context/ITestContextOptions.ts_
  - rename `callerRecursionDepth` to `callerMatchDepth`
- _src/Context/TestContext.ts_
  - default `callerMatchDepth` as 2
  - `defaultOptions` and `updateOptions`
  - `getTestGuid`
  - `getCurrentTestIdentifier`
- TestContext tests
- Fix for #6 
- Integration Tests
- RunSettings for pipeline

image

### CODE REVIEW

File: src/Constants.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        format(str: string, ...params: Array<any>): string;
+    }
+}
+
+String.format = (str: string, ...params: Array<string>): string => {
+    const regex = /{[0-9]+}/g;
+    const matches: Array<string | number> = str.match(regex);
+
+    matches.forEach((value: string, i, arr) => {
+        arr[i] = parseInt(value.substr(1, value.length - 2));
```
Comment (acesiddhu) [inline src/Constants.ts]: >value.substr(1, value.length - 2) [](start = 26, length = 33)

write tests!!!!!!!!

**Comment by acesiddhu** [comment]
>         if (fs.existsSync(attachmentPath)) {

invert

---
Refers to: src/Context/Attachments.ts:55 in 45736ea. [](commit_id = 45736eadaa400cc8ff7d107e174975abf4bdb0f9, deletion_comment = False)

**Comment by acesiddhu** [comment]
>         if (fs.existsSync(attachmentPath)) {

existsync already does lstat

---
Refers to: src/Context/Attachments.ts:55 in 45736ea. [](commit_id = 45736eadaa400cc8ff7d107e174975abf4bdb0f9, deletion_comment = False)

File: src/Context/TestContext.ts [hunk 2]
```diff
@@ -9,26 +12,29 @@ declare var it: any;
...
+    // tslint:disable-next-line:variable-name
+    export const Attachments = _Attachments;
+
     let itOverrideSuccess = false;
-    const itList = [];
     let options: ITestContextOptions;
+    
+    const itList = [];
+    const defaultOptions = <ITestContextOptions> {
+        callerMatchDepth: 0
```
Comment (acesiddhu) [inline src/Context/TestContext.ts]: >callerMatchDepth [](start = 8, length = 16)

make this atleast 1

File: src/Context/TestContext.ts [hunk 3]
```diff
@@ -52,35 +58,31 @@ export namespace TestContext {
...
         } catch (e) {
-            // tslint:disable-next-line:max-line-length
-            assert.fail(`jstestcontext: Could not get current test name with caller recursion depth ${options.callerRecursionDepth}. Refer to https://github.com/karanjitsingh/jstestcontext for correct usage.`);
+            assert.fail(String.format(Constants.CouldNotGetTestMethodError, options.callerMatchDepth));
             return false;
         }
     }
+    
+    function getTestName(caller: Function, getIdentifier: boolean = false) {
+        if (caller === Attachments.getTestAttachmentDirectory) {
```
Comment (acesiddhu) [inline src/Context/TestContext.ts]: >getTestAttachmentDirectory [](start = 35, length = 26)

club

File: test/Unit/StringFormatTests.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    });
+
+    it('Replace patterns correctly', () => {
+        const template = 'template {0} {1} template';
+        assert.equal(String.format(template, 'a', 'b', 'c'), 'template a b template');
+        assert.equal(String.format(template, 'a', 'b'), 'template a b template');
+        assert.equal(String.format(template, 'a'), 'template a  template');
+    });
+    
+    it('Will work if param is same as pattern', () => {
```
Comment (acesiddhu) [inline test/Unit/StringFormatTests.ts]: have a test case withut {}

Comment (acesiddhu) [reply test/Unit/StringFormatTests.ts]: and one with only { and one with only }

---
In reply to: [256706917](https://github.com/karanjitsingh/jstestcontext/pull/7#discussion_r256706917) [](ancestors = 256706917)

Comment (karanjitsingh) [reply test/Unit/StringFormatTests.ts]: it seems you don't trust regex :P

Comment (acesiddhu) [reply test/Unit/StringFormatTests.ts]: do i trust bcl, do i trust node do i trust ur code. if everything is on trust then we dont need tests :D

---
In reply to: [256707247](https://github.com/karanjitsingh/jstestcontext/pull/7#discussion_r256707247) [](ancestors = 256707247)

Comment (karanjitsingh) [reply test/Unit/StringFormatTests.ts]: updated

**Pull request merged**